copyspecial.py

```diff
@@ -16,29 +16,60 @@
 import argparse
 
 # This is to help coaches and graders identify student assignments
-__author__ = "???"
+__author__ = "knmarvel with madarp"
 
 
-# +++your code here+++
-# Write functions and modify main() to call them
+def get_special_paths(dir):
+    """takes a directory and returns a list of the absolute
+    paths of the special files in the given directory"""
+    spec_paths = []
+    for root, dirs, files in os.walk(dir):
+        for name in files:
+            if re.search("__(.+?)__", name):
+                if os.path.join(os.path.abspath(dir), name) not in spec_paths:
+                    spec_paths.append(os.path.join(os.path.abspath(dir), name))
+                else:
+                    return "ERROR"
+    return spec_paths
 
-def main():
-    # This snippet will help you get started with the argparse module.
+
+def copy_to(paths, dir):
+    """given a list of the paths, copy those files into the
+    given directory"""
+    for file in paths:
+        shutil.copy(file, dir)
+
+
+def zip_to(paths, dir):
+    """given a list of paths, zip those files up into the
+    given zipfile"""
+    cmd = ["zip", "-j", dir] + paths
+    subprocess.call(cmd)
+
+
+def parsing():
+    """parses arguments given as parameters in calling function"""
     parser = argparse.ArgumentParser()
+    parser.add_argument('--fromdir', default='./')
     parser.add_argument('--todir', help='dest dir for special files')
     parser.add_argument('--tozip', help='dest zipfile for special files')
-    # TODO need an argument to pick up 'from_dir'
-    args = parser.parse_args()
+    return parser.parse_args()
 
-    # TODO you must write your own code to get the cmdline args.
-    # Read the docs and examples for the argparse module about how to do this.
 
-    # Parsing command line arguments is a must-have skill.
-    # This is input data validation.  If something is wrong (or missing) with any
-    # required args, the general rule is to print a usage message and exit(1).
+def main():
+    """Calls all functions to return above listed results"""
+    args = parsing()
+    spec_paths = get_special_paths(args.fromdir)
+    if spec_paths == "ERROR":
+        print("Error: duplicate special files.")
```

Comment:
I would put the error messages in with the function that's creating them.  I wouldn't have thought to go looking down in main for an error happening elsewhere...